pkg/kube/client_test.go

```diff
@@ -254,32 +256,79 @@ func TestBuild(t *testing.T) {
 }
 
 func TestPerform(t *testing.T) {
+	var fn func(info *resource.Info) error
```

Comment:
I don't think it should be added to this place.
You should add a fake response in TestClient

https://github.com/helm/helm/blob/b1e24764325135b364a7f81fb0aab683d190c546/pkg/kube/client_test.go#L121

Reply:
Hi,

Do you mean TestBuild?
~~~
func TestBuild(t *testing.T) {
        tests := []struct {
                name      string
                namespace string
                reader    io.Reader
                count     int
                err       bool
        }{
                {
                        name:      "Valid input",
                        namespace: "test",
                        reader:    strings.NewReader(guestbookManifest),
                        count:     6,
                }, {
                        name:      "Valid input, deploying resources into different namespaces",
                        namespace: "test",
                        reader:    strings.NewReader(namespacedGuestbookManifest),
                        count:     1,
                },
        }

        c := newTestClient(t)
        for _, tt := range tests {
                t.Run(tt.name, func(t *testing.T) {
                        // Test for an invalid manifest
                        infos, err := c.Build(tt.reader, false)
                        if err != nil && !tt.err {
                                t.Errorf("Got error message when no error should have occurred: %v", err)
                        } else if err != nil && strings.Contains(err.Error(), "--validate=false") {
                                t.Error("error message was not scrubbed")
                        }

                        if len(infos) != tt.count {
                                t.Errorf("expected %d result objects, got %d", tt.count, len(infos))
                        }
                })
        }
}
~~~

There is no function `TestClient`.

Reply:
I understand what you mean, now. perform only catches batch creation / deletion. While create / delete call perform, update does not call it. So in the UpdateTest, I cannot create my unit test (because the logic which I added to perform would not catch this issue).  Update calls createResource(info) directly, whereas Create calls perform(infolist, createResource), the same for delete. But this would mean that I have to create a wrapper function for the 3 atomic actions: create, update, delete, correct?

Reply:
alternatively, update could call "perform" ;-)

Reply:
I pushed a new suggestion. That should cover what you meant (so, retry create, update, delete on conflict) by creating wrapper functions for each of the original methods. I also adjusted the unit test. I yet have to test this in my environment to see if this really fixes the issue that I'm hitting, but it should. Any suggestions to make this prettier are more than welcome. Also, are there any other methods which would need to be wrapped?

Reply:
Sorry, please forgive me for bad English, so that there is no expression clear

I just mean it to recommend that we build an error response in fakehttpClient, used to do unit testing.

like this
```
Client: fake.CreateHTTPClient(func(req *http.Request) (*http.Response, error) {
			if req.Method == http.MethodPost {}
}
```

Reply:
Sorry, please forgive me for bad English, so that there is no expression clear

I just mean it to recommend that we build an error response in fakehttpClient, used to do unit testing.

like this
```
Client: fake.CreateHTTPClient(func(req *http.Request) (*http.Response, error) {
}
```

Reply:
Hi,

Please see now. I moved the retry logic from "perform" to `createResource` and `deleteResource`. So, when `Update` calls these methods, it will benefit from the retry, too.

`updateResource` is more complex and I did not dare change it. Particularly because the issue which I found is on `createResource`. Therefore, I did not modify the `updateResource` method and it does not retry at the moment. 

I added a 409 Conflict in `TestUpdate` and I created a new test method `TestCreate` which runs 2 different tests. 

I also found DATA RACEs while testing when running `TestCreate` with a list that contains more than 1 pod. That might be an issue with tests or with the existing code (`perform`/`batchPerform` call 2 levels of nested go functions), it's not part of anything new that I added.

Please let me know what you think.

- Andreas